### ISSUE TITLE
TypeScript: missing, wrong and refactoring types

### ISSUE THREAD

**Issue by wellwelwel** [sidorares/node-mysql2#2072]
> Please, don't use this area to indicate new Issues, give preference to [opening a new Issue](https://github.com/sidorares/node-mysql2/issues/new).

### Pool
- [x] Missing Connection URI overload in `mysql` import, like `createPool('uri')`
  - [#2084](https://github.com/sidorares/node-mysql2/pull/2084)
- [x] Confusing `maxIdle` option description
  - [#2070](https://github.com/sidorares/node-mysql2/issues/2070)
  - [#1858](https://github.com/sidorares/node-mysql2/pull/1858)
  - [#2073](https://github.com/sidorares/node-mysql2/pull/2073)

### Pool Cluster
> It doesn't have any tests yet ⚠️
- [x] Missing Connection URI overload, like `poolCluster.add('conn', 'uri')`
  - [#2084](https://github.com/sidorares/node-mysql2/pull/2084)
- [x] Missing config option `url` in `add` method
  - [#1991](https://github.com/sidorares/node-mysql2/issues/1991)
  - [#2084](https://github.com/sidorares/node-mysql2/pull/2084)
- [x] `end`, `of` and `add` methods can't be in a `PoolCluster` connection
  - [#2084](https://github.com/sidorares/node-mysql2/pull/2084)
- [x] `of` type can't be a `PoolCluster`
  - [#1991](https://github.com/sidorares/node-mysql2/issues/1991)
  - [#2036](https://github.com/sidorares/node-mysql2/pull/2036)
  - [#2084](https://github.com/sidorares/node-mysql2/pull/2084)
- [x] Pool Cluster is missing in `promise.d.ts`
  - [#2035](https://github.com/sidorares/node-mysql2/pull/2035)
  - [#2084](https://github.com/sidorares/node-mysql2/pull/2084)

### Constants
- [x] Missing export for `Types` (it's necessary to fix both **JS** and **TS**)
  - [#1938](https://github.com/sidorares/node-mysql2/issues/1938)
  - [#2084](https://github.com/sidorares/node-mysql2/pull/2084)

<hr>

### Notes

- All fixes need to be covered by tests ([see practical examples here](https://github.com/sidorares/node-mysql2/blob/fefe036512624b384d48c926b11e7a986f7ae57c/test/tsc-build/index.ts))
- If new issues are identified, they will be added here

<hr>

<details open>

<summary>
  <b>🔬 About TypeScript Tests</b>
</summary>

<br>

The tests are based on TypeScript compilation and not executes the compiled code.
>  **Reason:** Given that the main codebase and its entire maintenance are carried out in JavaScript, creating tests in TypeScript could potentially hinder development in its primary language. It allows the development of the main driver and its types in parallel.
> See: #1957 and https://github.com/sidorares/node-mysql2/pull/1995#issuecomment-1539268223.

</details>

<hr>

<details>

<summary>
  <b>ℹ️ About refactoring</b>
</summary>

<br>

The **PR** #2067 solves all the following refactoring, but it has become too long and hard to review. Then, I plan to split it into smaller tasks and submit them again.

> The idea behind the refactoring revolves around:
> - Organizing and patterning the code for better readability and maintenance
> - Improving the performance of the code editor when compiling the types in its display
> - Increasing compatibility across multiple TypeScript configurations (`tsconfig`)

</details>

<hr>

## 1) Small refactoring
- [x] Include all `*.ts` files into `tsconfig.json` (except for tests)
  - [#2073](https://github.com/sidorares/node-mysql2/pull/2073)
- [x] Resolve **TS** `eslint` errors
  - [#2073](https://github.com/sidorares/node-mysql2/pull/2073)
- [x] Remove unused **TS** tests
  - [#2073](https://github.com/sidorares/node-mysql2/pull/2073)
- [x] Remove `@types/mocha`, `@types/chai`, `mocha` and `chai` dependencies
  - [#2073](https://github.com/sidorares/node-mysql2/pull/2073)
- [x] Then, include every `*.ts` file to `lint` script (including the tests)
  - [#2073](https://github.com/sidorares/node-mysql2/pull/2073)

## 2) Large refactoring
- [x] Pattern all modules importing and exporting
  - [#2073](https://github.com/sidorares/node-mysql2/pull/2073)
- [x] Join spread out types
  - [#2075](https://github.com/sidorares/node-mysql2/pull/2075)
- [x] Create a reusable way for `query` and `execute` overloads
  - [#2075](https://github.com/sidorares/node-mysql2/pull/2075)

## 3) Optional refactoring
- [x] Increasing compatibility with `moduleResolution: nodenext`
  - [#2073](https://github.com/sidorares/node-mysql2/pull/2073)
- [x] Remove `strictFunctionTypes: false` from `tsconfig.json`
  - [#2073](https://github.com/sidorares/node-mysql2/pull/2073)
<hr>

<details>

<summary>
  <b>🕐 Open Issues that have possibly already been resolved</b>
</summary>

<br>

- #1761
- #1816
- #1936

</details>

**Comment by wellwelwel** [comment]
@sidorares, I took the liberty of pinning this Issue, feel free to change it and this content.
